File: leetcode-in-python3/solution/104_maximum_depth_of_binary_tree.py
from datastruct.treenode import TreeNode
from collections import deque
import random
import unittest


class Solution:

    def maxDepth(self, root: TreeNode) -> int:
        num = random.randint(0, 1)
        d = {
            0: self.dfs,
            1: self.postorder,
            2: self.bfs
        }
        return d[num](root)

    def postorder(self, node: TreeNode) -> int:
        if node is None:
            return 0

        ans = 0
        deep = 1

        stack = []
        prev_node = None

        while True:
            while node is not None:
                stack.append((node, deep))
                deep += 1
                node = node.left

            if len(stack) <= 0:
                break

            if stack[-1][0].right != prev_node:
                node = stack[-1][0].right
                prev_node = None
            else:
                prev_node, deep = stack.pop()
                ans = max(ans, deep)
        return ans

    def bfs(self, root: TreeNode) -> int:
        if root is None:
            return 0
        deep = 0
        queue = deque()
        queue.append(root)
        while len(queue) > 0:
            deep += 1
            size = len(queue)
            for i in range(size):
                node = queue.popleft()
                if node.left is not None:
                    queue.append(node.left)
                if node.right is not None:
                    queue.append(node.right)
        return deep

    def dfs(self, root: TreeNode) -> int:
        def _dfs(node: TreeNode) -> int:
            if node is None:
                return 0
            return 1 + max(_dfs(node.left), _dfs(node.right))
        return _dfs(root)


class TestSolution(unittest.TestCase):

    def setUp(self):
        self.test_case = [
            ([3, 9, 20, None, None, 15, 7], 3),
            ([], 0),
        ]
        self.s = Solution()

    def test_solution(self):
        for nums, answer in self.test_case:
            root = TreeNode.create(nums)
            ans = self.s.maxDepth(root)
            self.assertEqual(answer, ans)


if __name__ == '__main__':
    unittest.main()

File: leetcode-in-rust/src/solution/_104_maximum_depth_of_binary_tree.rs
use crate::datastruct::treenode::TreeNode;

pub struct Solution {}

use std::rc::Rc;
use std::cell::RefCell;
use std::cmp::max;

impl Solution {
    pub fn max_depth(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        fn dfs(node: &Option<Rc<RefCell<TreeNode>>>) -> i32 {
            // if let Some(node) = node {
            //     let node = node.borrow();
            //     return 1 + max(
            //         dfs(&node.left),
            //         dfs(&node.right),
            //     );
            // }
            // 0
            //
            match node {
                Some(node) => {
                    let node = node.borrow();
                    1 + max(dfs(&node.left), dfs(&node.right))
                }
                None => 0,
            }
        }
        dfs(&root)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_solution() {
        let test_case = vec![
            (vec![Some(3), Some(9), Some(20), None, None, Some(15), Some(7)], 3),
        ];
        for (nums, answer) in test_case {
            let root = TreeNode::create(nums);
            let ans = Solution::max_depth(root);
            assert_eq!(ans, answer)
        }
    }
}
